Model/CategoryString/Builder.php

```diff
@@ -0,0 +1,126 @@
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\CategoryString;
+
+use Magento\Catalog\Api\CategoryRepositoryInterface;
+use Magento\Catalog\Model\Category;
+use Magento\Store\Model\Store;
+use Magento\Catalog\Model\Product;
+use Magento\Framework\Event\ManagerInterface;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Nosto\Tagging\Logger\Logger as NostoLogger;
+
+class Builder
+{
+    private $logger;
+    private $categoryRepository;
+    private $eventManager;
+
+    /**
+     * @param CategoryRepositoryInterface $categoryRepository
+     * @param NostoLogger $logger
+     * @param ManagerInterface $eventManager
+     */
+    public function __construct(
+        CategoryRepositoryInterface $categoryRepository,
+        NostoLogger $logger,
+        ManagerInterface $eventManager
+    ) {
+        $this->categoryRepository = $categoryRepository;
+        $this->logger = $logger;
+        $this->eventManager = $eventManager;
+    }
+
+    /**
+     * @param Product $product
+     * @param Store $store
+     * @return array
+     */
+    public function buildCategories(Product $product, Store $store)
+    {
+        $categories = [];
+        foreach ($product->getCategoryCollection() as $category) {
+            $categoryString = $this->build($category, $store);
+            if (!empty($categoryString)) {
+                $categories[] = $categoryString;
+            }
+        }
+
+        return $categories;
+    }
+
+    /**
+     * @param Category $category
+     * @param Store $store
+     * @return null|string
+     */
+    public function build(Category $category, Store $store)
+    {
+        $nostoCategory = '';
+        try {
+            $data = [];
+            $path = $category->getPath();
+            $storeId = $store->getId();
+            foreach (explode('/', $path) as $categoryId) {
+                try {
+                    $category = $this->categoryRepository->get($categoryId, $storeId);
+                } catch (NoSuchEntityException $noSuchEntityException) {
+                    continue;
+                }
+                if ($category instanceof Category
+                    && $category->getLevel() > 1
+                    && !empty($category->getName())
+                ) {
+                    $data[] = $category->getName();
+                }
+            }
+            $nostoCategory = count($data) ? '/' . implode('/', $data) : '';
+        } catch (\Exception $e) {
+            $this->logger->exception($e);
+        }
+        if (empty($nostoCategory)) {
+            $nostoCategory = null;
+        } else {
+            $this->eventManager->dispatch(
+                'nosto_category_string_load_after',
+                ['category' => $nostoCategory, 'magentoCategory' => $category]
```

Comment:
Rename "category" to "categoryString" to avoid confusion with the other category builder
```suggestion
                ['categoryString' => $nostoCategory, 'magentoCategory' => $category]
```